### DISCUSSION TITLE
Adding GenAI support to Redash

### DISCUSSION THREAD

**Discussion by imohitmayank** [getredash/redash#6733] [Ideas]
Just testing the waters to check if adding Generative AI-based features to Redash is under consideration!? Here are some ideas, 
- **Auto-generate Individual Queries**: As Redash has access to tbl and their schema, it should be possible to auto-generate queries based on either the title or comment of the query or provided prompt.
- **Auto-complete Individual Queries**: Given a partial query and its comment or prompt, it should be possible to complete the query.
- **Suggest Title for Queries**: Based on the queries in the Query page, we can suggest title. 
- **Add comments**: User can select part or complete queries and AI can add comments for better explainability.
- **Error Debugging**: Based on the executed query and the error, AI can suggest more verbose error mitigation steps.
- **Edit existing queries**: User can select part or complete queries, and provide prompt to modify the query. Ex: "add one more case to this query" or "simplify the query"

And all of these functionalities could be supported by connecting to OpenAI (or other) 3rd party GenAI service provider. User can set their OpenAI keys in the Settings page.

Please share your thoughts and add more ideas if I am missing anything. Thanks!

**Comment by guidopetri** [comment]
I don't think this would be safe from an opsec standpoint. OpenAI would get to know the schema of all of the data available to the user. The other suggestions (title, error help) don't seem like they'd be worth the effort of implementing, to be honest.

**Reply by imohitmayank** [reply]
This could be an optional setting so only interested users can opt-in. OpenAI supports an opt-out policy, so you can ask them to not use your data. Also, this is a list of potential ideas and it might make sense for different use cases, for example people just getting started with SQL might find it useful to have AI powered error-debugging feature.

**Reply by gabrieldutra** [reply]
I agree with this not being an opsec concern, lots of other products are embedding Azure OpenAI for some of these for example. It would be opt-in for sure, as each org would configure their key only when they want to have those features (would consider it by org, not by user because of the 3rd party data though).

**Comment by patcon** [comment]
Example of how flow works for a browser extension intended for Metabsae: https://avanty.app/

**Reply by imohitmayank** [reply]
This is good, thanks for sharing. Added some missing ideas to the original list.

**Reply by zachliu** [reply]
have you tried the copycat's https://github.com/DeepInsight-AI/DeepBI?
they just gave redash a face-lift and added gen-ai support :joy_cat:
